In [1]:
import numpy as np
import pandas as pd

In [2]:
url = r'C:\Users\nuria\Proyecto_EDA\data\raw\Student Depression Dataset.csv'

### 1. Inspeccionar el tamaño y estructura de los datos

In [20]:
df = pd.read_csv(r'C:\Users\nuria\Proyecto_EDA\data\raw\Student Depression Dataset.csv')
df
# CGPA: Promedio de calificaciones u otros puntajes académicos.

id  Gender   Age           City Profession  Academic Pressure  \
0           2    Male  33.0  Visakhapatnam    Student                5.0   
1           8  Female  24.0      Bangalore    Student                2.0   
2          26    Male  31.0       Srinagar    Student                3.0   
3          30  Female  28.0       Varanasi    Student                3.0   
4          32  Female  25.0         Jaipur    Student                4.0   
...       ...     ...   ...            ...        ...                ...   
27896  140685  Female  27.0          Surat    Student                5.0   
27897  140686    Male  27.0       Ludhiana    Student                2.0   
27898  140689    Male  31.0      Faridabad    Student                3.0   
27899  140690  Female  18.0       Ludhiana    Student                5.0   
27900  140699    Male  27.0          Patna    Student                4.0   

       Work Pressure  CGPA  Study Satisfaction  Job Satisfaction  \
0                0.0  8.97                 2.0               0.0   
1                0.0  5.90                 5.0               0.0   
2                0.0  7.03                 5.0               0.0   
3                0.0  5.59                 2.0               0.0   
4                0.0  8.13                 3.0               0.0   
...              ...   ...                 ...               ...   
27896            0.0  5.75                 5.0               0.0   
27897            0.0  9.40                 3.0               0.0   
27898            0.0  6.61                 4.0               0.0   
27899            0.0  6.88                 2.0               0.0   
27900            0.0  9.24                 1.0               0.0   

          Sleep Duration Dietary Habits    Degree  \
0              5-6 hours        Healthy   B.Pharm   
1              5-6 hours       Moderate       BSc   
2      Less than 5 hours        Healthy        BA   
3              7-8 hours       Moderate       BCA   
4              5-6 hours       Moderate    M.Tech   
...                  ...            ...       ...   
27896          5-6 hours      Unhealthy  Class 12   
27897  Less than 5 hours        Healthy       MSc   
27898          5-6 hours      Unhealthy        MD   
27899  Less than 5 hours        Healthy  Class 12   
27900  Less than 5 hours        Healthy       BCA   

      Have you ever had suicidal thoughts ?  Work/Study Hours  \
0                                       Yes               3.0   
1                                        No               3.0   
2                                        No               9.0   
3                                       Yes               4.0   
4                                       Yes               1.0   
...                                     ...               ...   
27896                                   Yes               7.0   
27897                                    No               0.0   
27898                                    No              12.0   
27899                                   Yes              10.0   
27900                                   Yes               2.0   

       Financial Stress Family History of Mental Illness  Depression  
0                   1.0                               No           1  
1                   2.0                              Yes           0  
2                   1.0                              Yes           0  
3                   5.0                              Yes           1  
4                   1.0                               No           0  
...                 ...                              ...         ...  
27896               1.0                              Yes           0  
27897               3.0                              Yes           0  
27898               2.0                               No           0  
27899               5.0                               No           1  
27900               3.0                              Yes           1  

[27901 row

In [19]:
df.head()

id  Gender   Age           City Profession  Academic Pressure  \
0   2    Male  33.0  Visakhapatnam    Student                5.0   
1   8  Female  24.0      Bangalore    Student                2.0   
2  26    Male  31.0       Srinagar    Student                3.0   
3  30  Female  28.0       Varanasi    Student                3.0   
4  32  Female  25.0         Jaipur    Student                4.0   

   Work Pressure  CGPA  Study Satisfaction  Job Satisfaction  \
0            0.0  8.97                 2.0               0.0   
1            0.0  5.90                 5.0               0.0   
2            0.0  7.03                 5.0               0.0   
3            0.0  5.59                 2.0               0.0   
4            0.0  8.13                 3.0               0.0   

      Sleep Duration Dietary Habits   Degree  \
0          5-6 hours        Healthy  B.Pharm   
1          5-6 hours       Moderate      BSc   
2  Less than 5 hours        Healthy       BA   
3          7-8 hours       Moderate      BCA   
4          5-6 hours       Moderate   M.Tech   

  Have you ever had suicidal thoughts ?  Work/Study Hours  Financial Stress  \
0                                   Yes               3.0               1.0   
1                                    No               3.0               2.0   
2                                    No               9.0               1.0   
3                                   Yes               4.0               5.0   
4                                   Yes               1.0               1.0   

  Family History of Mental Illness  Depression  
0                               No           1  
1                              Yes           0  
2                              Yes           0  
3                              Yes           1  
4                               No           0

In [4]:
df.tail()

id  Gender   Age       City Profession  Academic Pressure  \
27896  140685  Female  27.0      Surat    Student                5.0   
27897  140686    Male  27.0   Ludhiana    Student                2.0   
27898  140689    Male  31.0  Faridabad    Student                3.0   
27899  140690  Female  18.0   Ludhiana    Student                5.0   
27900  140699    Male  27.0      Patna    Student                4.0   

       Work Pressure  CGPA  Study Satisfaction  Job Satisfaction  \
27896            0.0  5.75                 5.0               0.0   
27897            0.0  9.40                 3.0               0.0   
27898            0.0  6.61                 4.0               0.0   
27899            0.0  6.88                 2.0               0.0   
27900            0.0  9.24                 1.0               0.0   

          Sleep Duration Dietary Habits    Degree  \
27896          5-6 hours      Unhealthy  Class 12   
27897  Less than 5 hours        Healthy       MSc   
27898          5-6 hours      Unhealthy        MD   
27899  Less than 5 hours        Healthy  Class 12   
27900  Less than 5 hours        Healthy       BCA   

      Have you ever had suicidal thoughts ?  Work/Study Hours  \
27896                                   Yes               7.0   
27897                                    No               0.0   
27898                                    No              12.0   
27899                                   Yes              10.0   
27900                                   Yes               2.0   

       Financial Stress Family History of Mental Illness  Depression  
27896               1.0                              Yes           0  
27897               3.0                              Yes           0  
27898               2.0                               No           0  
27899               5.0                               No           1  
27900               3.0                              Yes           1

In [22]:
df.shape

(27901, 18)

In [45]:
df.set_index('id')

Gender   Age           City Profession  Academic Pressure  \
id                                                                  
2         Male  33.0  Visakhapatnam    Student                5.0   
8       Female  24.0      Bangalore    Student                2.0   
26        Male  31.0       Srinagar    Student                3.0   
30      Female  28.0       Varanasi    Student                3.0   
32      Female  25.0         Jaipur    Student                4.0   
...        ...   ...            ...        ...                ...   
140685  Female  27.0          Surat    Student                5.0   
140686    Male  27.0       Ludhiana    Student                2.0   
140689    Male  31.0      Faridabad    Student                3.0   
140690  Female  18.0       Ludhiana    Student                5.0   
140699    Male  27.0          Patna    Student                4.0   

        Work Pressure  CGPA  Study Satisfaction  Job Satisfaction  \
id                                                                  
2                 0.0  8.97                 2.0               0.0   
8                 0.0  5.90                 5.0               0.0   
26                0.0  7.03                 5.0               0.0   
30                0.0  5.59                 2.0               0.0   
32                0.0  8.13                 3.0               0.0   
...               ...   ...                 ...               ...   
140685            0.0  5.75                 5.0               0.0   
140686            0.0  9.40                 3.0               0.0   
140689            0.0  6.61                 4.0               0.0   
140690            0.0  6.88                 2.0               0.0   
140699            0.0  9.24                 1.0               0.0   

           Sleep Duration Dietary Habits    Degree  \
id                                                   
2               5-6 hours        Healthy   B.Pharm   
8               5-6 hours       Moderate       BSc   
26      Less than 5 hours        Healthy        BA   
30              7-8 hours       Moderate       BCA   
32              5-6 hours       Moderate    M.Tech   
...                   ...            ...       ...   
140685          5-6 hours      Unhealthy  Class 12   
140686  Less than 5 hours        Healthy       MSc   
140689          5-6 hours      Unhealthy        MD   
140690  Less than 5 hours        Healthy  Class 12   
140699  Less than 5 hours        Healthy       BCA   

       Have you ever had suicidal thoughts ?  Work/Study Hours  \
id                                                               
2                                        Yes               3.0   
8                                         No               3.0   
26                                        No               9.0   
30                                       Yes               4.0   
32                                       Yes               1.0   
...                                      ...               ...   
140685                                   Yes               7.0   
140686                                    No               0.0   
140689                                    No              12.0   
140690                                   Yes              10.0   
140699                                   Yes               2.0   

        Financial Stress Family History of Mental Illness  Depression Country  
id                                                                             
2                    1.0                               No           1     NaN  
8                    2.0                              Yes           0     NaN  
26                   1.0                              Yes           0     NaN  
30                   5.0                              Yes           1     NaN  
32                   1.0                               No           0     NaN  
...                  ...                              ...         ...     ...  
140685    

Este conjunto de datos contiene 18 variables (columnas) y 27901 filas

In [46]:
# Verificar si hay jerarquías o estructuras anidadas
# Comprobar si hay listas o diccionarios en alguna columna
for column in df.columns:
    if df[column].apply(lambda x: isinstance(x, (list, dict))).any():
        print(f"La columna '{column}' tiene estructuras anidadas.")

No hay jerarquías ni estructuras anidadas

### 2. Verificar tipos de datos y nombres de columnas

In [47]:
df.columns

Index(['id', 'Gender', 'Age', 'City', 'Profession', 'Academic Pressure',
       'Work Pressure', 'CGPA', 'Study Satisfaction', 'Job Satisfaction',
       'Sleep Duration', 'Dietary Habits', 'Degree',
       'Have you ever had suicidal thoughts ?', 'Work/Study Hours',
       'Financial Stress', 'Family History of Mental Illness', 'Depression',
       'Country'],
      dtype='object')

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27901 entries, 0 to 27900
Data columns (total 19 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   id                                     27901 non-null  int64  
 1   Gender                                 27901 non-null  object 
 2   Age                                    27901 non-null  float64
 3   City                                   27901 non-null  object 
 4   Profession                             27901 non-null  object 
 5   Academic Pressure                      27901 non-null  float64
 6   Work Pressure                          27901 non-null  float64
 7   CGPA                                   27901 non-null  float64
 8   Study Satisfaction                     27901 non-null  float64
 9   Job Satisfaction                       27901 non-null  float64
 10  Sleep Duration                         27901 non-null  object 
 11  Di

No hay valores nulos

|Columna| Descripción| Tipo de dato|
|---|---|---|
|id|id|int|
|Gender|Genero|str|
|Age|Edad|int|
|City|Ciudad|str|
|Profession|Profesión|str|
|Academic Pressure|Presión académica|float|
|Work Pressure|Presión laboral|float|
|CGPA|Promedio de calificaciones u otros puntajes académicos|float|
|Study Satisfaction|Satisfacción de estudio|float|
|Job Satisfaction|Satisfacción de trabajo|float|
|Sleep Duration|Duración de sueño|str|
|Dietary Habits|Dieta|str|
|Degree|Año de estudios|str|
|Have you ever had suicidal thoughts?|¿Alguna vez has tenidos pensamientos suicidas?|str|
|Work/Study Hours|Horas de trabajo/estudio|float|
|Financial Stress|Estres economico|float|
|Family History of Mental Illness|Antecedentes familiares de enfermedad mental|str|
|Depression|Depresión|int|

Obsevaciones
- Memoria: 3.8+ MB
- La edad sale como flotante
- Sleep durantion sale tipo object
- Todos los textos salen tipo object pero llaman la atención 'Gender' y 'Have you ever had suicidal thoughts?'

### 3. Identificar valores duplicados o errores evidentes

In [50]:
duplicados = df.duplicated()
duplicados.head(10)

0    False
1    False
2    False
3    False
4    False
5    False
6    False
7    False
8    False
9    False
dtype: bool

In [51]:
# Observamos que tenga valores missing
df[df.duplicated(keep = False)] 

Empty DataFrame
Columns: [id, Gender, Age, City, Profession, Academic Pressure, Work Pressure, CGPA, Study Satisfaction, Job Satisfaction, Sleep Duration, Dietary Habits, Degree, Have you ever had suicidal thoughts ?, Work/Study Hours, Financial Stress, Family History of Mental Illness, Depression, Country]
Index: []

No tenemos valores duplicados exactos ni valores missing

In [52]:
df.isnull().sum()

id                                           0
Gender                                       0
Age                                          0
City                                         0
Profession                                   0
Academic Pressure                            0
Work Pressure                                0
CGPA                                         0
Study Satisfaction                           0
Job Satisfaction                             0
Sleep Duration                               0
Dietary Habits                               0
Degree                                       0
Have you ever had suicidal thoughts ?        0
Work/Study Hours                             0
Financial Stress                             3
Family History of Mental Illness             0
Depression                                   0
Country                                  27901
dtype: int64

In [53]:
df.nunique()

id                                       27901
Gender                                       2
Age                                         34
City                                        52
Profession                                  14
Academic Pressure                            6
Work Pressure                                3
CGPA                                       332
Study Satisfaction                           6
Job Satisfaction                             5
Sleep Duration                               5
Dietary Habits                               4
Degree                                      28
Have you ever had suicidal thoughts ?        2
Work/Study Hours                            13
Financial Stress                             5
Family History of Mental Illness             2
Depression                                   2
Country                                      0
dtype: int64

In [54]:
df.groupby('City')

In [55]:
city_counts = df.groupby('City').size()
print(city_counts)


City
3.0                      1
Agra                  1094
Ahmedabad              951
Bangalore              767
Bhavna                   2
Bhopal                 934
Chennai                885
City                     2
Delhi                  768
Faridabad              461
Gaurav                   1
Ghaziabad              745
Harsh                    1
Harsha                   2
Hyderabad             1340
Indore                 643
Jaipur                1036
Kalyan                1570
Kanpur                 609
Khaziabad                1
Kibara                   1
Kolkata               1066
Less Delhi               1
Less than 5 Kalyan       1
Lucknow               1155
Ludhiana              1111
M.Com                    1
M.Tech                   1
ME                       1
Meerut                 825
Mihir                    1
Mira                     1
Mumbai                 699
Nagpur                 651
Nalini                   1
Nalyan                   1
Nandini                

In [56]:
city_means = df.groupby('City')['Depression'].mean()
print(city_means)
# Calcula la media de otras columnas agrupandolas por cuidad

City
3.0                   1.000000
Agra                  0.534735
Ahmedabad             0.672976
Bangalore             0.608866
Bhavna                1.000000
Bhopal                0.619914
Chennai               0.596610
City                  0.500000
Delhi                 0.609375
Faridabad             0.587852
Gaurav                1.000000
Ghaziabad             0.574497
Harsh                 1.000000
Harsha                0.500000
Hyderabad             0.669403
Indore                0.604977
Jaipur                0.524131
Kalyan                0.594904
Kanpur                0.541872
Khaziabad             1.000000
Kibara                0.000000
Kolkata               0.608818
Less Delhi            0.000000
Less than 5 Kalyan    0.000000
Lucknow               0.554978
Ludhiana              0.582358
M.Com                 1.000000
M.Tech                0.000000
ME                    0.000000
Meerut                0.597576
Mihir                 1.000000
Mira                  1.000000
Mum

## Observaciones
- El conjunto de datos se compone de 18 columnas (variables) y 27901 filas
- Utiliza un espacio de memoria de 3.5+ MB
- Solo hay valores nulos en 'Financial Stress'
- No hay valores duplicados
- No hay datos faltantes
- Hay algunos tipos de dato que probablemente tenga que cambiar
- No hay jerarquías o estructuras anidadas

In [63]:
df = df.drop(columns=['Job Satisfaction'])
print(df.head())

KeyError: "['Job Satisfaction'] not found in axis"

In [65]:
df = df.drop(columns=['Work Pressure'])

In [66]:
df.head(2)

id  Gender   Age           City Profession  Academic Pressure  CGPA  \
0   2    Male  33.0  Visakhapatnam    Student                5.0  8.97   
1   8  Female  24.0      Bangalore    Student                2.0  5.90   

   Study Satisfaction Sleep Duration Dietary Habits   Degree  \
0                 2.0      5-6 hours        Healthy  B.Pharm   
1                 5.0      5-6 hours       Moderate      BSc   

  Have you ever had suicidal thoughts ?  Work/Study Hours  Financial Stress  \
0                                   Yes               3.0               1.0   
1                                    No               3.0               2.0   

  Family History of Mental Illness  Depression Country  
0                               No           1     NaN  
1                              Yes           0     NaN

In [68]:
# Guarda el dataframe en una archivo CSV nuevo (sin las columnas eliminadas)
df.to_csv(r'C:\Users\nuria\Proyecto_EDA\data\raw\Student Depression Dataset sin job satisfaction.csv', index='id')

In [69]:
df

id  Gender   Age           City Profession  Academic Pressure  \
0           2    Male  33.0  Visakhapatnam    Student                5.0   
1           8  Female  24.0      Bangalore    Student                2.0   
2          26    Male  31.0       Srinagar    Student                3.0   
3          30  Female  28.0       Varanasi    Student                3.0   
4          32  Female  25.0         Jaipur    Student                4.0   
...       ...     ...   ...            ...        ...                ...   
27896  140685  Female  27.0          Surat    Student                5.0   
27897  140686    Male  27.0       Ludhiana    Student                2.0   
27898  140689    Male  31.0      Faridabad    Student                3.0   
27899  140690  Female  18.0       Ludhiana    Student                5.0   
27900  140699    Male  27.0          Patna    Student                4.0   

       CGPA  Study Satisfaction     Sleep Duration Dietary Habits    Degree  \
0      8.97                 2.0          5-6 hours        Healthy   B.Pharm   
1      5.90                 5.0          5-6 hours       Moderate       BSc   
2      7.03                 5.0  Less than 5 hours        Healthy        BA   
3      5.59                 2.0          7-8 hours       Moderate       BCA   
4      8.13                 3.0          5-6 hours       Moderate    M.Tech   
...     ...                 ...                ...            ...       ...   
27896  5.75                 5.0          5-6 hours      Unhealthy  Class 12   
27897  9.40                 3.0  Less than 5 hours        Healthy       MSc   
27898  6.61                 4.0          5-6 hours      Unhealthy        MD   
27899  6.88                 2.0  Less than 5 hours        Healthy  Class 12   
27900  9.24                 1.0  Less than 5 hours        Healthy       BCA   

      Have you ever had suicidal thoughts ?  Work/Study Hours  \
0                                       Yes               3.0   
1                                        No               3.0   
2                                        No               9.0   
3                                       Yes               4.0   
4                                       Yes               1.0   
...                                     ...               ...   
27896                                   Yes               7.0   
27897                                    No               0.0   
27898                                    No              12.0   
27899                                   Yes              10.0   
27900                                   Yes               2.0   

       Financial Stress Family History of Mental Illness  Depression Country  
0                   1.0                               No           1     NaN  
1                   2.0                              Yes           0     NaN  
2                   1.0                              Yes           0     NaN  
3                   5.0                              Yes           1     NaN  
4                   1.0                               No           0     NaN  
...                 ...                              ...         ...     ...  
27896               1.0                              Yes           0     NaN  
27897               3.0                              Yes           0     NaN  
27898               2.0                               No           0     NaN  
27899               5.0                               No           1     NaN  
27900               3.0                              Yes           1     NaN  

[27901 rows x 17 columns]